// @widget-lab/platform-connectors/out/strategies/BaseStrategy.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { requirejsPromise } from "@widget-lab/3ddashboard-utils";
import SecurityContextsNotEnabledError from "../errors/SecurityContextsNotEnabledError";
import { ERROR_PASSPORT_DISCONNECTED } from "../utils/constants";
export default class BaseStrategy {
  constructor({
    allowTenantsSelection = false,
    allowSecurityContextSelection = false,
    compassURL = undefined,
    securityContexts = undefined
  }) {
    _defineProperty(this, "loadAvailableSecurityContexts", async () => {
      if (this.userDefinedSecurityContexts) {
        return this.userDefinedSecurityContexts.map(sc => ({
          label: sc,
          value: sc
        }));
      }

      const collaborationSpaces = await this.loadAvailableCollaborationSpaces();
      this.securityContextsList = collaborationSpaces.reduce((acc, collabspace) => {
        return acc.concat(collabspace.couples.map(couple => ({
          collaborationSpace: collabspace.name,
          organization: couple.organization,
          role: couple.role
        })));
      }, []);
      return this.securityContextsList.map(sc => ({
        label: `${sc.role.nls || sc.role.name} | ${sc.organization.name} | ${sc.collaborationSpace}`,
        value: `${sc.role.name}.${sc.organization.name}.${sc.collaborationSpace}`
      }));
    });

    _defineProperty(this, "loadAvailableCollaborationSpaces", async () => {
      const currentTenantId = await this.getCurrentTenantId();
      const servicePath = `/resources/modeler/pno/person?current=true&select=collabspaces&select=firstname&select=lastname&tenant=${currentTenantId}`;
      const url3DSpace = await this.getCurrentTenantServiceUrl("3DSpace");
      const WAFData = await requirejsPromise("DS/WAFData/WAFData");
      return new Promise((resolve, reject) => {
        WAFData.authenticatedRequest(`${url3DSpace}${servicePath}`, {
          type: "json",
          onComplete: data => {
            resolve(data.collabspaces);
          },
          onFailure: () => {
            reject("Error while retrieving security contexts. If running in Run your App mode, please make sure both 3DSpace URL and tenant are correct.");
          },
          onPassportError: () => {
            reject(ERROR_PASSPORT_DISCONNECTED);
          }
        });
      });
    });

    _defineProperty(this, "getSecurityContext", () => {
      if (this.disableSecurityContexts) {
        return Promise.resolve(undefined);
      }

      if (!this.securityContextPreferenceManager) {
        throw new SecurityContextsNotEnabledError();
      }

      return this.securityContextPreferenceManager.readValue();
    });

    this.allowTenantsSelection = allowTenantsSelection;
    this.allowSecurityContextSelection = allowSecurityContextSelection;
    this.securityContextsList = null;
    this.securityContextsLoadingPromise = null;
    this.securityContextPreferenceManager = null;
    this.tenantSelectionPreferenceManager = null;
    this.tenantsLoadingPromise = null;
    this.servicesLoadingPromise = null;
    this.tenantsList = [];
    this.compassURL = compassURL;
    this.userDefinedSecurityContexts = securityContexts;
  }

}
