// @widget-lab/platform-connectors/out/strategies/TrustedStrategy.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import BaseStrategy from "./BaseStrategy";
import TenantSelectionPreferenceManager from "../preferences/common/TenantPreferenceManager";
import SecurityContextPreferenceManager from "../preferences/common/SecurityContextPreferenceManager";
import { getPlatforms } from "../utils/getPlatforms";
import { requirejsPromise, widget } from "@widget-lab/3ddashboard-utils";
export default class TrustedStrategy extends BaseStrategy {
  constructor(options) {
    super(options);

    _defineProperty(this, "createPreferences", () => {
      this.tenantSelectionPreferenceManager = new TenantSelectionPreferenceManager();
      this.securityContextPreferenceManager = new SecurityContextPreferenceManager();
    });

    _defineProperty(this, "resetSecurityContext", (dispatchEvent = false) => {
      if (!this.resetSecurityContextIsLoading) {
        this.resetSecurityContextIsLoading = true;
        this.securityContextsLoadingPromise = this.loadAvailableSecurityContexts().then(securityContextsListOptions => {
          this.securityContextPreferenceManager.setSecurityContextsListOptions(securityContextsListOptions, this.allowSecurityContextSelection);
        }).catch(err => {
          throw err;
        }).finally(() => {
          // use boolean because "onTenantUpdate" event is fired twice a row
          this.resetSecurityContextIsLoading = false; // dispatch onEdit event to update UI preferences

          if (dispatchEvent) widget.dispatchEvent("onEdit");
        });
      }
    });

    _defineProperty(this, "run", () => {
      this.tenantsLoadingPromise = this.loadAvailableTenants().then(tenants => {
        this.tenantsList = tenants;
        const tenantsListOptions = tenants.map(t => ({
          label: t.displayName,
          value: t.platformId
        }));
        this.tenantSelectionPreferenceManager.setTenantsListOptions(tenantsListOptions, this.allowTenantsSelection); // then load all services

        const currentTenant = this.tenantsList.find(t => t.platformId === this.tenantSelectionPreferenceManager.readValue());

        if (currentTenant && currentTenant["3DCompass"]) {
          this.servicesLoadingPromise = getPlatforms(currentTenant["3DCompass"]).then(tenantsWithAllServices => {
            this.tenantsList.forEach(tenant => {
              var _tenantsWithAllServic;

              tenant.allServices = ((_tenantsWithAllServic = tenantsWithAllServices.find(t => t.id === tenant.platformId)) === null || _tenantsWithAllServic === void 0 ? void 0 : _tenantsWithAllServic.services) || [];
            });
          });
        }
      });

      try {
        this.resetSecurityContext();
      } catch (err) {
        throw new Error(err);
      }

      widget.addEvent("onTenantUpdate", () => {
        this.resetSecurityContext(true);
      });
      return Promise.all([this.tenantsLoadingPromise, this.servicesLoadingPromise, this.securityContextsLoadingPromise]);
    });

    _defineProperty(this, "getCurrentTenantId", async () => {
      await this.tenantsLoadingPromise;
      return this.tenantSelectionPreferenceManager.readValue();
    });

    _defineProperty(this, "getCurrentTenant", async () => {
      await this.tenantsLoadingPromise;
      let tenant = this.tenantsList[0];
      const tenantId = this.tenantSelectionPreferenceManager.readValue();
      tenant = this.tenantsList.find(t => t.platformId === tenantId);
      const {
        displayName,
        platformId,
        allServices
      } = tenant;
      return {
        services: allServices,
        name: displayName,
        id: platformId
      };
    });

    _defineProperty(this, "getCurrentTenantServiceUrl", async serviceName => {
      var _tenant$services$find;

      await this.tenantsLoadingPromise;
      await this.servicesLoadingPromise;
      const tenant = await this.getCurrentTenant();
      serviceName = serviceName.toLowerCase();
      return (_tenant$services$find = tenant.services.find(s => s.id.toLowerCase() === serviceName || s.name.toLowerCase() === serviceName)) === null || _tenant$services$find === void 0 ? void 0 : _tenant$services$find.url;
    });

    _defineProperty(this, "getCurrentTenantServices", async () => {
      await this.tenantsLoadingPromise;
      await this.servicesLoadingPromise;
      const tenant = await this.getCurrentTenant();
      return tenant.services;
    });

    _defineProperty(this, "loadAvailableTenants", () => {
      return requirejsPromise("DS/i3DXCompassServices/i3DXCompassServices").then(i3DXCompassServices => {
        const GET_SERVICES_TIMEOUT = 10000;
        return new Promise((resolve, reject) => {
          const timeout = setTimeout(() => {
            reject();
          }, GET_SERVICES_TIMEOUT);
          i3DXCompassServices.getPlatformServices({
            // OOTB function does not have a onFailure function...
            onComplete: results => {
              clearTimeout(timeout);
              resolve(results);
            }
          });
        });
      });
    });
  }

}
