// @widget-lab/platform-connectors/out/strategies/UntrustedStrategy.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import BaseStrategy from "./BaseStrategy";
import TenantSelectionPreferenceManager from "../preferences/common/TenantPreferenceManager";
import Manual3DCompassPreferenceManager from "../preferences/untrustedMode/Manual3DCompassPreferenceManager";
import SecurityContextPreferenceManager from "../preferences/common/SecurityContextPreferenceManager";
import { getPlatforms } from "../utils/getPlatforms";
import { findCompassUrl } from "../utils/utils";
import { widget } from "@widget-lab/3ddashboard-utils";
export default class UntrustedStrategy extends BaseStrategy {
  constructor(options) {
    super(options);

    _defineProperty(this, "createPreferences", () => {
      this.tenantSelectionPreferenceManager = new TenantSelectionPreferenceManager();
      this.securityContextPreferenceManager = new SecurityContextPreferenceManager();
      this.servicesPreferencesManagers = {};
      let defaultCompassUrl = this.compassURL;

      if (!this.compassURL) {
        // autodetect compassURL if not set by default in initializer
        defaultCompassUrl = findCompassUrl();
      }

      this.servicesPreferencesManagers["3DCompass"] = new Manual3DCompassPreferenceManager(defaultCompassUrl);
    });

    _defineProperty(this, "run", async () => {
      this.tenantsLoadingPromise = await this.fillTenantsList();

      if (this.tenantsList.length > 0) {
        this.securityContextsLoadingPromise = this.fillSecurityContextsList();
      }

      let resetTenantAndSecurityContext = async (resetTenant = false, resetSecurityContext = false) => {
        if (!this.resetTenantAndSecurityContextIsLoading) {
          // use this boolean because "on3DCompassURLUpdate" event is fired two time a row
          this.resetTenantAndSecurityContextIsLoading = true; // fill the tenants via 3DCompass URL if there is no tenant already or if resetTenant is true

          if (this.tenantsList.length === 0 || resetTenant) {
            this.tenantsLoadingPromise = null;
            this.tenantsLoadingPromise = this.fillTenantsList();
          } // fill security context if we need it and only if tenants are loaded or if resetSecurityContext is true


          if (this.tenantsList.length !== 0 || resetSecurityContext) {
            this.securityContextsLoadingPromise = null;
            this.securityContextsLoadingPromise = await this.fillSecurityContextsList();
          }
        }

        this.resetTenantAndSecurityContextIsLoading = false;
      };

      widget.addEvent("onRefresh", () => {
        resetTenantAndSecurityContext();
      });
      widget.addEvent("on3DCompassURLUpdate", () => {
        resetTenantAndSecurityContext(true, true).catch(err => {
          throw err;
        }).finally(() => widget.dispatchEvent("onEdit")); // dispatch onEdit event to update UI preferences
      });
      widget.addEvent("onTenantUpdate", () => {
        resetTenantAndSecurityContext(false, true).catch(err => {
          throw err;
        }).finally(() => widget.dispatchEvent("onEdit")); // dispatch onEdit event to update UI preferences
      }); // returning Promise.all for a single promise to avoid checking if this.allowSecurityContextSelection is null.

      return Promise.all([this.tenantsLoadingPromise, this.securityContextsLoadingPromise]);
    });

    _defineProperty(this, "fillSecurityContextsList", async () => {
      try {
        const securityContextsListOptions = await this.loadAvailableSecurityContexts();
        this.securityContextPreferenceManager.setSecurityContextsListOptions(securityContextsListOptions, this.allowSecurityContextSelection);
      } catch (err) {
        this.securityContextPreferenceManager.setSecurityContextsListOptions([], this.allowSecurityContextSelection);
        throw err;
      }
    });

    _defineProperty(this, "fillTenantsList", async () => {
      try {
        const tenants = await this.loadAvailableTenants();
        this.tenantsList = tenants;
        const tenantsListOptions = tenants.map(t => ({
          label: t.name,
          value: t.id
        }));
        this.tenantSelectionPreferenceManager.setTenantsListOptions(tenantsListOptions, this.allowTenantsSelection);
        return tenants;
      } catch (err) {
        this.tenantSelectionPreferenceManager.setTenantsListOptions([], this.allowTenantsSelection);
        throw err;
      }
    });

    _defineProperty(this, "loadAvailableTenants", async () => {
      // get 3DCompass preference value
      const compassUrl = this.servicesPreferencesManagers["3DCompass"].readValue();
      if (compassUrl === "") throw new Error("3DCompass url is not set in widget preferences. On cloud, it's easy to determine ; for example if the platform is https://eu1-ifwe.3dexperience.3ds.com then the 3DCompass base URL is https://eu1-compass.3dexperience.3ds.com/enovia. On Premise, 3DCompass URL is the same as 3DSpace URL"); // do authenticated request with 3DCompass url to retrieve all tenants

      const tenants = await getPlatforms(compassUrl);
      return tenants;
    });

    _defineProperty(this, "getCurrentTenantId", async () => {
      await this.tenantsLoadingPromise;
      return this.tenantSelectionPreferenceManager.readValue();
    });

    _defineProperty(this, "getCurrentTenant", async () => {
      await this.tenantsLoadingPromise;
      return this.tenantsList.find(t => t.id === this.tenantSelectionPreferenceManager.readValue());
    });

    _defineProperty(this, "getCurrentTenantServiceUrl", async serviceName => {
      var _this$tenantsList$fin;

      await this.tenantsLoadingPromise;
      const tenantId = await this.getCurrentTenantId();
      serviceName = serviceName.toLowerCase();
      return (_this$tenantsList$fin = this.tenantsList.find(x => x.id === tenantId).services.find(s => s.id.toLowerCase() === serviceName || s.name.toLowerCase() === serviceName)) === null || _this$tenantsList$fin === void 0 ? void 0 : _this$tenantsList$fin.url;
    });

    _defineProperty(this, "getCurrentTenantServices", async () => {
      var _this$tenantsList$fin2;

      await this.tenantsLoadingPromise;
      const tenantId = await this.getCurrentTenantId();
      return (_this$tenantsList$fin2 = this.tenantsList.find(x => x.id === tenantId)) === null || _this$tenantsList$fin2 === void 0 ? void 0 : _this$tenantsList$fin2.services;
    });
  }

}
