// @widget-lab/platform-connectors/out/index.js
import PlatformInitializer from "./PlatformInitializer";
import PlatformConnectorsNotInitializedError from "./errors/PlatformConnectorsNotInitializedError";
import PlatformConnectorsAlreadyInitializedError from "./errors/PlatformConnectorsAlreadyInitializedError";
let platformInitializer = null;
export const initPlatformConnectors = async options => {
  if (platformInitializer) {
    throw new PlatformConnectorsAlreadyInitializedError();
  }

  platformInitializer = new PlatformInitializer();
  return await platformInitializer.init(options);
};
export const call3DSwym = (url, options) => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getInterface3DSwym().then(interface3DSwym => {
    return interface3DSwym.call3DSwym(url, options);
  });
};
export const call3DCompass = (url, options) => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getInterface3DCompass().then(interface3DCompass => {
    return interface3DCompass.call3DCompass(url, options);
  });
};
export const call3DSpace = (url, options) => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getInterface3DSpace().then(interface3DSpace => {
    return interface3DSpace.call3DSpace(url, options);
  });
};
export const call3DWebServices = (serviceName, webServicePath, options) => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getInterfaceGeneric().then(interfaceGeneric => {
    return interfaceGeneric.call3DWebServices(serviceName, webServicePath, options);
  });
};
export const get3DSpaceServiceUrl = () => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getInterface3DSpace().then(interface3DSpace => {
    return interface3DSpace.get3DSpaceServiceUrl();
  });
};
export const getCurrentTenantServiceUrl = serviceName => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getStrategy().then(strategy => {
    return strategy.getCurrentTenantServiceUrl(serviceName);
  });
};
export const getCurrentTenantServices = () => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getStrategy().then(strategy => {
    return strategy.getCurrentTenantServices();
  });
};
export const getCurrentTenantId = () => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getStrategy().then(strategy => {
    return strategy.getCurrentTenantId();
  });
};
export const getCurrentSecurityContext = () => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getStrategy().then(strategy => {
    return strategy.getSecurityContext();
  });
};
export const getCurrentTenant = () => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getStrategy().then(strategy => {
    return strategy.getCurrentTenant();
  });
};
export const get3DSpaceCsrfToken = () => {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  return platformInitializer.getInterface3DSpace().then(interface3DSpace => {
    return interface3DSpace.loadCsrfToken();
  });
};
export async function getTransientTicket() {
  if (!platformInitializer) {
    throw new PlatformConnectorsNotInitializedError();
  }

  const interfaceGeneric = await platformInitializer.getInterfaceGeneric();
  const data = await interfaceGeneric.call3DWebServices("3dpassport", "/api/authenticated/cas/transient", {
    method: "GET",
    type: "json"
  });
  return data.access_token;
}
export * from "./utils/constants";
