// @widget-lab/platform-connectors/out/PlatformInitializer.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import TrustedStrategy from "./strategies/TrustedStrategy";
import InterfaceGeneric from "./services/InterfaceGeneric";
import Interface3DSwym from "./services/Interface3DSwym";
import Interface3DSpace from "./services/Interface3DSpace";
import Interface3DCompass from "./services/Interface3DCompass";
import UntrustedStrategy from "./strategies/UntrustedStrategy";
import { isTrusted } from "@widget-lab/3ddashboard-utils";
export default class PlatformInitializer {
  constructor() {
    _defineProperty(this, "init", async options => {
      this.initPromise = this.initStrategy(options);
      this.strategy = await this.initPromise;
      this.initServices(this.strategy);
      return this.initPromise;
    });

    _defineProperty(this, "initStrategy", async options => {
      const isRunYourApp = !isTrusted();
      const strategy = isRunYourApp ? new UntrustedStrategy(options) : new TrustedStrategy(options);
      strategy.createPreferences();
      await strategy.run();
      return strategy;
    });

    _defineProperty(this, "initServices", strategy => {
      this.interfaceGeneric = new InterfaceGeneric(strategy);
      this.interface3DSwym = new Interface3DSwym(strategy);
      this.interface3DSpace = new Interface3DSpace(strategy);
      this.interface3DCompass = new Interface3DCompass(strategy);
    });

    _defineProperty(this, "getInterfaceGeneric", async () => {
      await this.initPromise;
      return this.interfaceGeneric;
    });

    _defineProperty(this, "getInterface3DSwym", async () => {
      await this.initPromise;
      return this.interface3DSwym;
    });

    _defineProperty(this, "getInterface3DCompass", async () => {
      await this.initPromise;
      return this.interface3DCompass;
    });

    _defineProperty(this, "getInterface3DSpace", async () => {
      await this.initPromise;
      return this.interface3DSpace;
    });

    _defineProperty(this, "getStrategy", async () => {
      await this.initPromise;
      return this.strategy;
    });

    this.strategy = null;
  }

}
