// @widget-lab/3ddashboard-utils/out/utils/index.js
import { STANDALONE_WIDGET_ID, Widget } from "./Widget";
import { RequireJSUtils } from "./RequirejsUtils";
class UWAUtils {
    constructor() {
    }
    log(args) {
        console.log(args);
    }
}
class VisibilityHandler {
    constructor() {
        this.pageVisible = true;
        this.bodyVisible = true;
        this.visible = true;
        this.callbacks = [];
    }
    fireChange(changes) {
        Object.assign(this, changes);
        const vis = this.pageVisible && this.bodyVisible;
        if (this.visible !== vis) {
            this.visible = vis;
            this.callbacks.forEach(cb => cb(this.visible));
        }
    }
    onChange(cb) {
        cb(this.visible);
        this.callbacks.push(cb);
    }
}
class Utils {
    static init() {
        document.addEventListener("visibilitychange", () => Utils.visibility.fireChange({ pageVisible: !document.hidden }));
        const observer = new window.IntersectionObserver(([elem]) => Utils.visibility.fireChange({ bodyVisible: elem.isIntersecting }), { threshold: 0.1 });
        const body = document.querySelector("body");
        if (body)
            observer.observe(body);
        if (!Utils._isWidget) {
            window.widget = new Widget();
            window.UWA = new UWAUtils();
            RequireJSUtils.init();
        }
        if (window.BASE_URL) {
            window.widget.uwaPath = window.BASE_URL;
        }
        else
            window.widget.uwaPath = window.widget.uwaUrl.substring(0, window.widget.uwaUrl.lastIndexOf("/") + 1);
        if (!Utils._isWidget) {
            window.widget.widgetDomain = window.widget.uwaPath;
        }
        try {
            __webpack_public_path__ = window.widget.uwaPath;
        }
        catch (e) {
        }
    }
    static onVisibilityChange(cb) {
        Utils.visibility.onChange(cb);
    }
    static disableDefaultCSS(disable) {
        let disableOptions = true;
        if (typeof disable === "boolean" && disable === false) {
            disableOptions = false;
        }
        const styleSheets = document.styleSheets;
        for (let i = 0; i < styleSheets.length; i++) {
            const sheet = styleSheets.item(i);
            for (const partialUrlToTest of Utils.widgetDefaultStyleSheets) {
                if (sheet && sheet.href && sheet.href.indexOf(partialUrlToTest) !== -1) {
                    sheet.disabled = disableOptions;
                }
            }
        }
    }
    static isWidget() {
        return widget.id !== STANDALONE_WIDGET_ID;
    }
    static isTrusted() {
        try {
            return isWidget() && !!window.parent.location.href;
        }
        catch {
            return false;
        }
    }
}
Utils.widgetDefaultStyleSheets = ["UWA/assets/css/iframe.css"];
Utils.visibility = new VisibilityHandler();
Utils._isWidget = !!window.UWA;
Utils.init();
const widget = window.widget;
const UWA = window.UWA;
const { requirejs } = window;
const { requirejsPromise } = RequireJSUtils;
const { disableDefaultCSS, onVisibilityChange, isWidget, isTrusted } = Utils;
export { UWA, widget };
export { disableDefaultCSS, onVisibilityChange, isWidget, isTrusted, requirejs, requirejsPromise };
export { UWAUtils, Widget };
export * from "./Widget";
