// @widget-lab/3ddashboard-utils/out/utils/Widget.js
export const STANDALONE_WIDGET_ID = "standalone";
export class Widget {
    constructor() {
        this.title = "";
        this.uwaUrl = "./";
        this.readOnly = false;
        this.id = STANDALONE_WIDGET_ID;
        this.data = {};
        this.setValue = (prefName, value) => {
            this.preferences[prefName].value = value;
            this.savePrefsLocalStorage();
        };
        this.events = {};
        this.title = "";
        this.uwaPath = "";
        this.widgetDomain = "";
        this.preferences = this.loadPreferences();
    }
    loadPreferences() {
        const prefsLocalStr = localStorage.getItem(Widget.PREFERENCE_NAME);
        let prefsLocal;
        if (prefsLocalStr && prefsLocalStr !== "null") {
            try {
                prefsLocal = JSON.parse(prefsLocalStr);
            }
            catch (err) {
                localStorage.setItem(Widget.PREFERENCE_NAME, JSON.stringify({}));
            }
        }
        else {
            prefsLocal = {};
            localStorage.setItem(Widget.PREFERENCE_NAME, JSON.stringify(prefsLocal));
        }
        return prefsLocal;
    }
    savePrefsLocalStorage() {
        localStorage.setItem(Widget.PREFERENCE_NAME, JSON.stringify(this.preferences));
    }
    addEvent(event, callback, bind, priority) {
        this.events[event] = callback;
        if (event === "onLoad") {
            if (document.readyState === "loading") {
                window.addEventListener("DOMContentLoaded", callback);
            }
            else {
                callback();
            }
        }
        return this;
    }
    dispatchEvent(event, args, bind) {
        const cb = this.events[event];
        if (!cb)
            return;
        if (bind)
            cb.bind(bind, args);
        else
            cb();
        return this;
    }
    addPreference(pref) {
        if (!this.preferences[pref.name]) {
            pref.value = pref.defaultValue;
            this.preferences[pref.name] = pref;
            this.savePrefsLocalStorage();
        }
    }
    getPreference(prefName) {
        return this.preferences[prefName];
    }
    getValue(prefName) {
        return this.preferences[prefName]?.value;
    }
    setTitle(title) {
        this.title = title;
        window.document.title = this.title;
    }
    getTitle() {
        return this.title;
    }
    setIcon(icon) {
        let link = document.querySelector("link[rel~='icon']");
        if (!link) {
            link = document.createElement("link");
            link.rel = "icon";
            document.getElementsByTagName("head")[0].appendChild(link);
        }
        link.href = icon;
    }
}
Widget.PREFERENCE_NAME = "_prefs_4_Widget_";
