// @widget-lab/platform-connectors/out/preferences/common/TenantPreferenceManager.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { TENANT_PREFERENCE_NAME, OOTB_CURRENT_TENANT_PREFERENCE_NAME, TENANT_PREFERENCE_LABEL, ERROR_LIST_OPTION } from "../../utils/constants";
import { widget } from "@widget-lab/3ddashboard-utils";
export default class TenantPreferenceManager {
  constructor() {// Nothing to do here, this pref is a list and needs to wait for tenants

    _defineProperty(this, "setTenantsListOptions", (listOptions = [], isVisible = false) => {
      this.tenantsListOptions = listOptions; // Store available options to check while reading value

      let tenantListPreference = {
        name: TENANT_PREFERENCE_NAME,
        label: TENANT_PREFERENCE_LABEL,
        type: isVisible ? "list" : "hidden",
        options: listOptions.length > 0 ? listOptions : ERROR_LIST_OPTION,
        onchange: "onTenantUpdate"
      };

      if (!widget.getPreference(TENANT_PREFERENCE_NAME)) {
        // Need a specific default value on creation
        const platformId = widget.getValue(OOTB_CURRENT_TENANT_PREFERENCE_NAME);
        tenantListPreference = Object.assign(tenantListPreference, {
          defaultValue: platformId
        });
      }

      widget.addPreference(tenantListPreference);
    });

    _defineProperty(this, "readValue", () => {
      let tenantPreference = widget.getValue(TENANT_PREFERENCE_NAME);

      if (!this.tenantsListOptions.find(x => x.value === tenantPreference)) {
        var _this$tenantsListOpti;

        tenantPreference = (_this$tenantsListOpti = this.tenantsListOptions[0]) === null || _this$tenantsListOpti === void 0 ? void 0 : _this$tenantsListOpti.value;
        widget.setValue(TENANT_PREFERENCE_NAME, tenantPreference);
      }

      return widget.getValue(TENANT_PREFERENCE_NAME);
    });
  }

}
