// @widget-lab/3ddashboard-utils/out/requirejsModules/i3DXCompassServices.js
import { requirejsPromise } from "../utils/index";
export class i3DXCompassServices {
    async retrievePlatformServices(compassURL) {
        if (!this.platformServices)
            this.platformServices = new Promise(async (resolve, reject) => {
                if (!compassURL)
                    throw new Error('3DCompass URL is undefined. Use "compassURL" in options parameter for standalone context.');
                const WAFData = await requirejsPromise("DS/WAFData/WAFData");
                WAFData.authenticatedRequest(`${compassURL}/resources/AppsMngt/api/v1/services`, {
                    type: "json",
                    onComplete: (data) => {
                        resolve(data.platforms);
                    },
                    onFailure: () => {
                        reject();
                    }
                });
            });
        return this.platformServices;
    }
    getPlatformServices(options) {
        void this.retrievePlatformServices(options.compassURL).then(platforms => {
            const ret = platforms
                .filter(p => (options.platformId ? p.id === options.platformId : true))
                .map(({ id, name, services }) => {
                const servicesUrl = services
                    .filter(s => s.public === true)
                    .reduce((acc, curr) => {
                    acc[curr.name] = curr.url;
                    return acc;
                }, {});
                return { platformId: id, displayName: name, ...servicesUrl };
            });
            options.onComplete(ret);
        });
    }
    getServiceUrl(options) {
        void this.retrievePlatformServices(options.compassURL).then(platforms => {
            const ret = platforms
                .filter(p => (options.platformId ? p.id === options.platformId : true))
                .map(({ id, services }) => {
                return { platformId: id, url: services.find(s => s.name === options.serviceName)?.url };
            });
            if (ret.length === 0)
                options.onComplete(undefined);
            else if (ret.length === 1)
                options.onComplete(ret[0].url);
            else
                options.onComplete(ret);
        });
    }
}
