// @widget-lab/platform-connectors/out/utils/Http.js
import { requirejsPromise } from "@widget-lab/3ddashboard-utils";
import { ERROR_PASSPORT_DISCONNECTED } from "./constants";
const DEFAULT_TIMEOUT = 25000;

const authenticatedRequest = async (url, // Important: parameter deleteType is important for non-json calls to succeed in the authenticatedRequest
{
  method = "GET",
  data = {},
  headers = {},
  type = "json",
  deleteType = false,
  timeout = DEFAULT_TIMEOUT,
  onTimeout,
  onProgress,
  returnHeader = false
} = {}) => {
  // Manipulate headers as a Headers class : this will work in both cases where input headers are provided as Record or as Headers
  let enhancedHeaders = new Headers(headers);

  if (type === "json" && !deleteType) {
    enhancedHeaders.set("Content-Type", "application/json");
    enhancedHeaders.set("Accept", "application/json");
  } // WAFData does not accept a Headers class: it has to be transformed as a Record


  const finalHeaders = {};
  Array.from(enhancedHeaders.entries()).forEach(([key, value]) => finalHeaders[key] = value);
  const WAFData = await requirejsPromise("DS/WAFData/WAFData");
  return new Promise((resolve, reject) => {
    // Build the proper data to send to authenticatedRequest() depending on the different cases
    let wafData = data; // Cases (typeof data === "string") or (data instanceof FormData) for example

    if (type === "json" && !deleteType && typeof data === "object" && !(data instanceof FormData)) {
      if (Object.keys(data).length > 0) wafData = JSON.stringify(data);else wafData = undefined;
    }

    const obj = {
      onComplete: (data, responseHeaders) => {
        if (!returnHeader) resolve(data);else resolve({
          responseData: data,
          responseHeaders
        });
      },
      onFailure: (error, backendresponse, responseHeaders) => {
        // eslint-disable-next-line camelcase
        reject({
          error,
          backendresponse,
          response_hdrs: responseHeaders
        });
      },
      onPassportError: () => {
        reject(ERROR_PASSPORT_DISCONNECTED);
      },
      method,
      headers: finalHeaders,
      type,
      data: wafData,
      timeout,
      onTimeout,
      onProgress
    }; // Important: "delete obj.type" will not have the same effect as "{type: undefined}" in the authenticatedRequest call!

    if (deleteType) delete obj.type;
    WAFData.authenticatedRequest(url, obj);
  });
};

export { authenticatedRequest };
